### PR TITLE
Shared redispatch for 5.0

### PR DESCRIPTION
**Pull request by gsychev**
https://github.com/emqx/emqx/pull/7807 but for 5.0

### CODE REVIEW

**Review by ieQu1 — DISMISSED** [review]

File: apps/emqx/src/emqx_session.erl [hunk 1]
```diff
@@ -808,13 +799,43 @@ replay(ClientInfo, Session = #session{inflight = Inflight}) ->
...
-    run_hook('session.takenover', [ClientInfo, info(Session)]);
 terminate(ClientInfo, Reason, Session) ->
+    run_terminate_hooks(ClientInfo, Reason, Session),
+    redispatch_shared_messages(Session),
+    ok.
+
+run_terminate_hooks(ClientInfo, discarded, Session) ->
+    run_hook('session.discarded', [ClientInfo, info(Session)]);
+run_terminate_hooks(ClientInfo, takeovered, Session) ->
+    run_hook('session.takeovered', [ClientInfo, info(Session)]);
```
Comment (thalesmg) [inline apps/emqx/src/emqx_session.erl:810]:
```suggestion
run_terminate_hooks(ClientInfo, takenover, Session) ->
    run_hook('session.takenover', [ClientInfo, info(Session)]);
```

Comment (ieQu1) [reply apps/emqx/src/emqx_session.erl:810]: While this is a reasonable change, it may affect 3rd party plugins?

**Review by ieQu1 — APPROVED** [review]

File: apps/emqx/src/emqx_session.erl (hunk 1 above)
Comment (thalesmg) [reply apps/emqx/src/emqx_session.erl:810]: The previous code already uses `takenover`:

https://github.com/emqx/emqx/pull/7860/files#diff-1dcb97f8eab360778099824b93b350440db45d1d601286f97ef6d4487711f475L813

Comment (ieQu1) [reply apps/emqx/src/emqx_session.erl:810]: Nice, thanks for the answer!

**Review by thalesmg — APPROVED** [review]

**Pull request merged**